Add support for KeyVault Secrets

conniey opened the pull request:
## What does this PR do?

- Adds support for KeyVault secrets.  users can add or get a secret.

## GitHub issue number?

## Checklist before merging
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-mcp/blob/main/CONTRIBUTING.md) on pull request process, code style, and testing.**
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message.  This means that previously merged commits do not appear in the history of the PR.  For more information on cleaning up the commits in your PR,  [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If it's a core feature, I have added thorough tests.
- [ ] If it's a noteworthy bug fix or new feature, I have added an entry in CHANGELOG.md with a link back to the PR or issue
- [ ] Have a team member run [live tests](https://github.com/Azure/azure-mcp/blob/main/CONTRIBUTING.md#live-tests):
   - [ ] Team Member: Inspect PR for security issues before queueing a test run
   - [ ] Team Member: Add this comment to the pr `/azp run azure - mcp` to start the run

jongio commented on README.md:
```diff
@@ -86,7 +86,7 @@ The Azure MCP Server provides tools for interacting with the following Azure ser
 - Query logs using KQL
 - List available tables
-- Configure monitoring options
+- Configure monitoring optionsqw
```
```suggestion
- Configure monitoring options
```

joshfree replied:
@hallipr can you add cspell to this repo like the azure-sdk repos have enabled?

jongio commented on src/Commands/KeyVault/Secret/SecretCreateCommand.cs:
```diff
@@ -0,0 +1,114 @@
+        Create a new secret in an Azure Key Vault. This command creates a secret with the specified name and value
+        in the given vault.
+
+        Required arguments:
+        - subscription
+        - vault-name
+        - secret-name
+        - secret-value
```
Do we need these?  I think we can put required in the argumentdefinitions to communicate that to the client

jongio commented on src/Commands/KeyVault/Secret/SecretGetCommand.cs:
```diff
@@ -0,0 +1,90 @@
+        Get a secret from an Azure Key Vault. This command retrieves and displays details
+        about a specific key in the specified vault.
+
+        Required arguments:
+        - subscription
+        - vault-name
+        - secret-name
```
Same here, might not need.

joshfree commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 ### Features Added
 
+- Support for getting and creating secrets from Key Vault.
```
add a link to this PR too

joshfree reviewed (approved):
left comments to be fixed

jongio commented on src/Commands/KeyVault/Secret/SecretCreateCommand.cs:
```diff
@@ -0,0 +1,114 @@
+    private readonly Option<string> _secretNameOption = ArgumentDefinitions.KeyVault.SecretName.ToOption();
+    private readonly Option<string> _secretValueOption = ArgumentDefinitions.KeyVault.SecretValue.ToOption();
+
+    protected override string GetCommandName() => "create";
```
Please rebase with main.  We update the BaseCommand interface. Make sure you add the title property as well and update the McpServerTool attribute with the title.

jongio commented:
The codebase just went through a major refactor which will impact this PR.  Please make sure you rebase this branch and update it to match the new design.

joshfree commented:
Tagging @vcolin7

joshfree commented:
@vcolin7 can you help regenerate this and get this merged before your OOF?

conniey commented:
> @vcolin7 can you help regenerate this and get this merged before your OOF?

I've fixed this PR and added the features missed. @vcolin7 , can you review this PR?

joshfree commented on tests/Commands/Server/ToolOperationsTest.cs:
```diff
@@ -4,7 +4,9 @@
 using System.Text.Json;
 using AzureMcp.Commands;
 using AzureMcp.Commands.Server;
+using AzureMcp.Services.Interfaces;
```
I think our linter catches things like that, right @hallipr?

conniey replied:
Yes, it catches unused imports.

conniey commented on src/Commands/CommandFactory.cs:
```diff
@@ -304,9 +305,15 @@ private void RegisterKeyVaultCommands()
         keys.AddCommand("get", new KeyVault.Key.KeyGetCommand(GetLogger<KeyVault.Key.KeyGetCommand>()));
         keys.AddCommand("create", new KeyVault.Key.KeyCreateCommand(GetLogger<KeyVault.Key.KeyCreateCommand>()));
+
+        secret.AddCommand("get", new KeyVault.Secret.SecretGetCommand(GetLogger<KeyVault.Secret.SecretGetCommand>(),
```
At the moment, I was only thinking of get secret. I am happy to put in "create" secret if you think it is valuable.

joshfree commented:
@conniey I know this is assigned to @vcolin7 but I'm wondering if this is ready to merge ahead of the Tuesday release?

conniey commented on src/Commands/KeyVault/Secret/SecretGetCommand.cs:
```diff
@@ -0,0 +1,83 @@
+    private const string _commandTitle = "Get Key Vault Secret";
+    private readonly ILogger<SecretGetCommand> _logger = logger;
+    private readonly IKeyVaultService _keyVaultService = keyVaultService;
+    private readonly Option<string> _vaultOption = OptionDefinitions.KeyVault.VaultName;
```
I had initially thought this but the benefits don't warrant all the code. ` KeyVaultCommand`  would have to be generic to support:  `SubscriptionCommand<T>`.  Or/and, each command would have to support a BaseKVOption class that has shared "vault name" option along with any other args it wants.  Then we have to upcast the actual returned instance when BindingOptions so it is the correct..

conniey commented on src/Commands/KeyVault/Secret/SecretGetCommand.cs:
```diff
@@ -0,0 +1,83 @@
+                options.RetryPolicy);
+
+            context.Response.Results = ResponseResult.Create(
+                new SecretGetCommandResult(options.SecretName!, result),
```
https://learn.microsoft.com/en-us/dotnet/api/azure.security.keyvault.secrets.secretclient.getsecretasync?view=azure-dotnet

The documentation suggests otherwise, but I also tested this.

conniey commented:
/azp run azure - mcp

joshfree reviewed (approved):
Reviewed after merge conflicts resolved
